src/devices/funkbus.c

```diff
@@ -0,0 +1,197 @@
+            result++;
+        }
+    }
+    return result & 0x01;
+}
+
+static int funkbus_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+
+    for (uint16_t row = 0; row < bitbuffer->num_rows; row++) {
+        uint16_t len = bitbuffer->bits_per_row[row];
+        uint8_t *bin  = bitbuffer->bb[row];
+
+        if (len < 48) {
+            return DECODE_ABORT_LENGTH;
+        }
+
+        uint8_t get_c = 0;
+#define get(x) \
+    get_data_lsb(bin, get_c, get_c + x - 1); \
+    get_c += x
+
+        funkbus_packet_t packet;
+        packet.typ    = get(4);
+        packet.subtyp = get(4);
+
+        // remote
+        if (packet.typ != 0x4 || packet.subtyp != 0x3) {
+            return DECODE_ABORT_EARLY;
+        }
+
+        packet.sn     = get(20);
+        packet.r1     = get(2);
+        packet.bat    = get(1);
+        packet.r2     = get(2);
+        packet.sw     = get(3);
+        packet.ch     = get(2);
+        packet.r3     = get(1);
+        packet.action = get(2);
+        packet.repeat = get(1);
+        packet.global = get(1);
+        packet.parity = get(1);
+        packet.check  = get(4); // how to calc?
+
+        if (packet.parity != calc_parity(bin, 43)) {
+            return DECODE_FAIL_MIC;
+        }
+
+        // calc id based on sn, ch + sw
+        uint64_t id = packet.sw + (packet.ch << 4) + (packet.sn << 8);
+
+        /* clang-format off */
+        data_t *data = data_make(
+                "model",        "",                DATA_STRING, "funkbus",
+                "typ",          "",                DATA_STRING, "remote",
+                "id",           "id",              DATA_INT, id,
+                "sn",           "serial number",   DATA_INT, packet.sn,
+                "bat",          "battery empty",   DATA_INT, packet.bat,
+                "sw",           "switch",          DATA_INT, packet.sw,
+                "ch",           "channel",         DATA_INT, packet.ch,
+                "action",       "action",          DATA_INT, packet.action,
+                "repeat",       "repeat",          DATA_INT, packet.repeat,
+                "global",       "global",          DATA_INT, packet.global,
+#ifdef DEBUG
+                "parity",       "parity",          DATA_INT, packet.parity,
+                "check",        "check",           DATA_INT, packet.check,
+#endif
+                "mic",          "integrity",       DATA_STRING, "PARITY",
+                NULL);
+        /* clang-format on */
+
+        decoder_output_data(decoder, data);
+    }
+
+#ifdef DEBUG
+    bitbuffer_debug(bitbuffer);
+#endif
+
+    return 1;
+}
+
+static char *output_fields[] = {
+        "model",
+        "typ",
+        "id",
+        "sn",
+        "bat",
+        "sw",
+        "ch",
+        "action",
+        "repeat",
+        "global",
+#ifdef DEBUG
+        "parity",
+        "check",
+#endif
+        "mic",
+        NULL,
+};
+
+r_device funkbus_remote = {
+        .name        = "Funkbus remote",
```

Comment:
This can be more descriptive, e.g. "Funkbus / Instafunk (Berker, Jira, Jung)"

Reply:
changed